File: CrossCompileApp/composeApp/src/commonMain/kotlin/com/androbrain/crosscompileapp/data/repository/BudgetRepositoryImpl.kt
package com.androbrain.crosscompileapp.data.repository

import com.androbrain.crosscompileapp.data.datasource.BudgetDataSource
import com.androbrain.crosscompileapp.data.datasource.BudgetEntity
import com.androbrain.crosscompileapp.data.mapping.toEntity
import com.androbrain.crosscompileapp.data.mapping.toModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

class BudgetRepositoryImpl(
    private val budgetDataSource: BudgetDataSource,
) : BudgetRepository {
    override suspend fun insert(budget: BudgetModel) = budgetDataSource.insert(budget.toEntity())

    override suspend fun delete(budget: BudgetModel) = budgetDataSource.delete(budget.toEntity())

    override fun getAll(): Flow<List<BudgetModel>> = budgetDataSource.getAll()
        .map { it.map(BudgetEntity::toModel) }
}

File: iOSApp/iOSApp/data/repository/BudgetRepositoryImpl.swift
//
//  BudgetRepositoryImpl.swift
//  iOSApp
//
//  Created by Michal Ankiersztajn on 14/05/2025.
//
import Foundation

class BudgetRepositoryImpl : BudgetRepository {
    let dataSource: BudgetDataSource
    
    init(dataSource: BudgetDataSource) {
        self.dataSource = dataSource
    }
    
    func getAll() -> [BudgetModel] {
        return dataSource.getAll().map {
            toModel(entity: $0)
        }
    }
    
    func addOrReplace(budget: BudgetModel) {
        dataSource.addOrReplace(budget: budget)
    }
    
    func delete(id: UUID) {
        dataSource.delete(id: id)
    }
    
    private func toModel(entity: BudgetEntity) -> BudgetModel {
        return BudgetModel(
            amount: entity.amount, content: entity.content!, id: entity.id!, isSpending: entity.isSpending
        )
    }
}
